os/kernel/pthread/pthread_keydelete.c

```diff
@@ -77,39 +78,34 @@
 
 int pthread_key_delete(pthread_key_t key)
 {
-	struct pthread_tcb_s *rtcb = (FAR struct pthread_tcb_s *)this_task();
-	struct task_group_s *group = rtcb->cmn.group;
-	struct pthread_key_s *cur_key;
+	struct tcb_s *rtcb = this_task();
+	struct task_group_s *group = rtcb->group;
+	struct pthread_tcb_s *ptcb;
+	struct join_s *pjoin;
 
-	DEBUGASSERT(group && (rtcb->cmn.flags & TCB_FLAG_TTYPE_MASK) == TCB_FLAG_TTYPE_PTHREAD);
+	DEBUGASSERT(group);
 
 	/* Check key validation */
 
-	if (key < PTHREAD_KEYS_MAX) {
+	if (key < PTHREAD_KEYS_MAX && group->tg_key[key] == KEY_INUSE) {
+		sched_lock();
 
-		/* Find key data */
+		/* Delete destructor */
 
-		cur_key = (struct pthread_key_s *)pthread_key_find(rtcb, key);
-		if (cur_key == NULL) {
+		group->tg_destructor[key] = NULL;
 
-			/* No Key in the list */
+		/* Delete key data from all of thread in this group */
 
-			return EINVAL;
+		for (pjoin = group->tg_joinhead; pjoin; pjoin = pjoin->next) {
```

Comment:
, we are accessing a shared resource "group" , but we are not taking care of concurrent access . We may need to consider to use sched_lock / unlock to protect this critical section ?

Reply:
I refer it from pthread_findjoininfo(). It does not lock / unlock. The sched_gettcb() is a similar function but it does not check either. Do I need to lock/unlock?